tests/rados/stretch_cluster.py

```diff
@@ -0,0 +1,472 @@
+
+
+def sort_osd_sites(all_osd_details: dict) -> tuple:
+    """
+    Sorts the OSD's present such that the weights on two sites remains the same
+    Args:
+        all_osd_details: dictionary of OSD's containing the details
+            eg : {'2': {'weight': 0.01459, 'state': 'up', 'name': 'osd.2'},
+                '7': {'weight': 0.01459, 'state': 'up', 'name': 'osd.7'}}
+
+    Returns: Tuple of lists, containing the OSD list for the 2 sites
+        eg : ([1, 2, 3, 4, 5], [6, 7, 8, 9, 0])
+
+    """
+    site_a_osds = []
+    site_b_osds = []
+    osd_list = [x for x in all_osd_details.keys()]
+
+    # distributing the OSD's into two sites such that both sites have equal weight
+    while len(osd_list) > 1:
+        site_a_osd = osd_list.pop()
+        if not all_osd_details[site_a_osd]["state"] == "up":
+            log.error(f"OSD : {site_a_osd} is not up")
+            break
+        flag = 0
+        for osd in osd_list:
+            if all_osd_details[osd]["state"] == "up":
+                if (
+                    all_osd_details[site_a_osd]["weight"]
+                    == all_osd_details[osd]["weight"]
+                ):
+                    osd_list.remove(osd)
+                    site_a_osds.append(site_a_osd)
+                    site_b_osds.append(osd)
+                    flag = 1
+                    break
+            else:
+                log.error(f"OSD : {osd} is not up")
+                osd_list.remove(osd)
+        if not flag:
+            log.error(f"no peer OSD for: {site_a_osd} found")
+    log.info(
+        f"Proposed Site-A OSD's : {site_a_osds}\nProposed Site-B OSD's : {site_b_osds}"
+    )
+    return site_a_osds, site_b_osds
+
+
+def set_osd_sites(
+    node: CephAdmin, site_a_osds: list, site_b_osds: list, all_osd_details: dict
+) -> bool:
+    """
+    Collects all the details about the OSD's present on the cluster and distrubutes them among the two sites
+    Args:
+        node: Cephadm node where the commands need to be executed
+        site_a_osds: list of OSD's to be added to site A
+        site_b_osds: list of OSD's to be added to site B
+        all_osd_details: dictionary of OSD's containing the details
+            eg : {'2': {'weight': 0.01459, 'state': 'up', 'name': 'osd.2'},
+                '7': {'weight': 0.01459, 'state': 'up', 'name': 'osd.7'}}
+
+    Returns: True -> pass, False -> fail
+    """
+    # adding the identified OSD's into the respective sites
+    try:
+        for osd in site_a_osds:
+            cmd = f"ceph osd crush move {all_osd_details[osd]['name']} host=host1-{osd} datacenter=site1"
+            node.shell([cmd])
+            # sleeping for 20 seconds for osd to be moved
+            time.sleep(20)
+
+        for osd in site_b_osds:
+            cmd = f"ceph osd crush move {all_osd_details[osd]['name']} host=host2-{osd} datacenter=site2"
+            node.shell([cmd])
+            # sleeping for 20 seconds for osd to be moved
+            time.sleep(20)
```

Comment:
The difference between the code is `site_{a,b}_osds` and host `prefix`... IMO, we can implement a single method and call the method with different values here?

Is there a way to check the OSD has moved without sleeping but by way of confirmation

Reply:
We can confirm that the OSD has moved out completely only when number of PG's on that OSD become 0.

All OSD's have different number of PG's. Once we move the OSD's into sites, the PG's on them will start to move to other OSD's. This takes some time. The sleep of 20 seconds is given so that the cluster will mark the OSD in or out, and begin the PG movement across OSD's.
``` # ceph osd df
ID  CLASS  WEIGHT   REWEIGHT  SIZE     RAW USE  DATA     OMAP     META     AVAIL    %USE  VAR   PGS  STATUS
 0    hdd  0.01900   1.00000   20 GiB  736 MiB  263 MiB      0 B  473 MiB   19 GiB  3.59  1.12   88      up
13    hdd  0.01500   1.00000   15 GiB  563 MiB  185 MiB      0 B  378 MiB   14 GiB  3.67  1.14   68      up
14    hdd  0.01199   1.00000   12 GiB  429 MiB  243 MiB    1 KiB  185 MiB   12 GiB  3.49  1.09   58      up
 2    hdd  0.00999   1.00000   10 GiB  450 MiB  142 MiB      0 B  308 MiB  9.6 GiB  4.40  1.37   50      up
 3    hdd  0.01500   1.00000   15 GiB  565 MiB  187 MiB      0 B  377 MiB   14 GiB  3.68  1.15   55      up
 4    hdd  0.01199   1.00000   12 GiB  386 MiB  124 MiB      0 B  263 MiB   12 GiB  3.14  0.98   52      up
 7    hdd  0.00999   1.00000   10 GiB  468 MiB  167 MiB      0 B  301 MiB  9.5 GiB  4.57  1.43   45      up
 8    hdd  0.01500   1.00000   15 GiB  525 MiB  176 MiB      0 B  349 MiB   14 GiB  3.42  1.07   65      up
 9    hdd  0.01199   1.00000   12 GiB  259 MiB  202 MiB    3 KiB   57 MiB   12 GiB  2.11  0.66   65      up
 1    hdd  0.01500   1.00000   15 GiB  604 MiB  205 MiB      0 B  399 MiB   14 GiB  3.93  1.23   63      up
10    hdd  0.01900   1.00000   20 GiB  719 MiB  256 MiB      0 B  462 MiB   19 GiB  3.51  1.10   86      up
11    hdd  0.01500   1.00000   15 GiB  517 MiB  166 MiB      0 B  352 MiB   14 GiB  3.37  1.05   64      up
12    hdd  0.00999   1.00000   10 GiB  475 MiB  156 MiB      0 B  319 MiB  9.5 GiB  4.64  1.45   46      up
15    hdd  0.00999   1.00000   10 GiB  482 MiB  163 MiB      0 B  318 MiB  9.5 GiB  4.70  1.47   56      up
18    hdd  0.01199   1.00000   12 GiB  482 MiB  177 MiB      0 B  304 MiB   12 GiB  3.92  1.22   73      up
19    hdd  0.01199   1.00000   12 GiB  448 MiB  175 MiB      0 B  273 MiB   12 GiB  3.65  1.14   49      up
20    hdd  0.01199   1.00000   12 GiB  277 MiB  194 MiB    2 KiB   83 MiB   12 GiB  2.25  0.70   53      up
 6    hdd  0.01500   1.00000   15 GiB  552 MiB  197 MiB      0 B  355 MiB   14 GiB  3.59  1.12   56      up
16    hdd  0.01500   1.00000   15 GiB  189 MiB   15 MiB      0 B  174 MiB   15 GiB  1.23  0.38    0      up
17    hdd  0.01199   1.00000   12 GiB  125 MiB   15 MiB      0 B  110 MiB   12 GiB  1.02  0.32    0      up
 5    hdd  0.01900   1.00000   20 GiB  232 MiB   15 MiB      0 B  217 MiB   20 GiB  1.13  0.35    0      up
                       TOTAL  289 GiB  9.3 GiB  3.3 GiB  8.7 KiB  5.9 GiB  280 GiB  3.20
MIN/MAX VAR: 0.32/1.47  STDDEV: 1.09
 ```